libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -510,3 +737,96 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
 		);
 
 }
+
+void ContractLevelChecker::checkAmbiguousOverrides(ContractDefinition const& _contract) const
+{
+	vector<FunctionInfo> baseFuncs;
+
+	for (ContractDefinition const* c: resolveBaseContracts(_contract))
+		for (pair<string, FunctionInfo> const& funcPair: getOverridableFunctions(c))
+		{
+			auto baseFunc = funcPair.second.funcDef;
+
+			auto baseSearchResult =
+				find_if(
+					baseFuncs.cbegin(),
+					baseFuncs.cend(),
+					[&] (auto const& f) {
+						return hasEqualNameAndParameters(*(f.funcDef), *baseFunc);
+					}
+				);
+
+			if (baseSearchResult == baseFuncs.cend())
+			{
+				baseFuncs.emplace_back(funcPair.second);
+				continue;
+			}
+			else if (baseSearchResult->contract == funcPair.second.contract)
+				continue;
+
+			auto const& defFuncs = _contract.definedFunctions();
+			auto const derivedSearchresult = find_if(
+				defFuncs.cbegin(),
+				defFuncs.cend(),
+				[&] (auto const& f) {
+					return hasEqualNameAndParameters(*f, *baseFunc);
+				}
+			);
+
+			if (derivedSearchresult != defFuncs.cend())
+				continue;
+
+			SecondarySourceLocation ssl;
+			ssl.append("First definition here: ", baseSearchResult->funcDef->location());
+			ssl.append("Second definition here: ", baseFunc->location());
+
+			m_errorReporter.typeError(
+				_contract.location(),
+				ssl,
+				"Ambiguous base function " +
+				baseFunc->name() +
+				"() must be overridden."
+			);
+		}
+}
+
+vector<ContractDefinition const*> ContractLevelChecker::resolveOverrideList(OverrideSpecifier const& _overrides) const
+{
+	std::vector<ASTPointer<UserDefinedTypeName>> const& overrides = _overrides.overrides();
+
+	vector<ContractDefinition const*> resolved{overrides.size(), nullptr};
+
+	transform(
+		overrides.begin(),
+		overrides.end(),
+		resolved.begin(),
+		[&] (ASTPointer<UserDefinedTypeName> const& _name) -> ContractDefinition const*
+		{
+			Declaration const* decl  = _name->annotation().referencedDeclaration;
+			solAssert(decl, "Expected declaration to be resolved.");
+
+			if (ContractDefinition const* _contract = dynamic_cast<decltype(_contract)>(decl))
+				return _contract;
+
+			TypeType const* actualTypeType = dynamic_cast<TypeType const*>(decl->type());
+
+			// If it's not an enum, contract or struct, it will be caught
```

Comment:
Where exactly?
Also, please either use a comment or a reason string in the assertion.